go/adbc/driver/bigquery/record_reader.go

```diff
@@ -366,4 +366,160 @@ func (e emptyArrowIterator) SerializedArrowSchema() []byte {
+	if l.done {
+		return nil, iterator.Done
+	}
+
+	const batchSize = 1000
+	rows := make([][]bigquery.Value, 0, batchSize)
+
+	for i := 0; i < batchSize; i++ {
+		var row []bigquery.Value
+		err := l.iter.Next(&row)
+		if err == iterator.Done {
+			l.done = true
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		rows = append(rows, row)
+	}
+
+	if len(rows) == 0 {
+		return nil, iterator.Done
+	}
+
+	batch, err := rowsToArrowRecordBatch(l.schema, rows, l.alloc)
+	if err != nil {
+		return nil, err
+	}
+	defer batch.Release()
+
+	var buf bytes.Buffer
+	writer := ipc.NewWriter(&buf, ipc.WithSchema(batch.Schema()), ipc.WithAllocator(l.alloc))
+	if err := writer.Write(batch); err != nil {
+		return nil, err
+	}
+	if err := writer.Close(); err != nil {
+		return nil, err
+	}
+
+	return &bigquery.ArrowRecordBatch{
+		Data: buf.Bytes(),
+	}, nil
+}
+
+func (l *RowBasedArrowIterator) Schema() bigquery.Schema {
+	return l.schema
+}
+
+func (l *RowBasedArrowIterator) SerializedArrowSchema() []byte {
+	fields := make([]arrow.Field, len(l.schema))
+	for i, field := range l.schema {
+		f, err := buildField(field, 0)
+		if err != nil {
+			log.Fatalf("Error building field %s: %v", field.Name, err)
+		}
+		fields[i] = f
+	}
+
+	arrowSchema := arrow.NewSchema(fields, nil)
+
+	var buf bytes.Buffer
+	_ = ipc.NewWriter(&buf, ipc.WithSchema(arrowSchema))
+
+	return buf.Bytes()
+}
+
+func rowsToArrowRecordBatch(schema bigquery.Schema, rows [][]bigquery.Value, alloc memory.Allocator) (arrow.Record, error) {
+	if len(rows) == 0 {
+		return nil, fmt.Errorf("no rows to convert")
+	}
+
+	// Build a schema
+	fields := make([]arrow.Field, len(schema))
+	for i, field := range schema {
+		f, err := buildField(field, 0)
+		if err != nil {
+			return nil, err
+		}
+		fields[i] = f
+	}
+	arrowSchema := arrow.NewSchema(fields, nil)
+
+	// Build arrays for columns
+	builders := make([]array.Builder, len(schema))
+	for i, field := range fields {
+		builders[i] = array.NewBuilder(alloc, field.Type)
+	}
+	defer func() {
+		for _, b := range builders {
+			b.Release()
+		}
+	}()
+
+	// Populate data
+	for _, row := range rows {
+		for colIdx, val := range row {
+			if val == nil {
+				builders[colIdx].AppendNull()
+				continue
+			}
+
+			switch builder := builders[colIdx].(type) {
+			case *array.Date32Builder:
+				// BigQuery returns civil.Date for DATE columns
+				if d, ok := val.(civil.Date); ok {
+					t := time.Date(d.Year, time.Month(d.Month), d.Day, 0, 0, 0, 0, time.UTC)
+					builder.Append(arrow.Date32FromTime(t))
+				} else if t, ok := val.(time.Time); ok {
+					builder.Append(arrow.Date32FromTime(t))
+				} else {
+					builder.AppendNull()
+				}
+			case *array.TimestampBuilder:
+				if ts, ok := val.(time.Time); ok {
+					builder.Append(arrow.Timestamp(ts.UnixMicro()))
+				} else {
+					builder.AppendNull()
+				}
+			// TODO: Add support for other types as needed
+			default:
+				if sb, ok := builder.(*array.StringBuilder); ok {
```

Comment:
Can we return an error here instead? Otherwise we will get mysterious issues on the Rust side.

It's really sad that we have to do this for BigQuery as well :/